### PR TITLE
std_pip_args: use PythonPipBuilder.std_args(...) instead

### PR DESCRIPTION
**Pull request by haampie**
Currently `std_pip_args` is set in `build_environment.py` for any package,
including ones that don't depend on python or pip.

This commit replaces globals with direct calls to `PythonPipBuilder.std_args`,
so (A) the origin is clear, and (B) the dependency on build_environment.py
is effectively removed, which was an input that was not reflected in the package
hash.

### CODE REVIEW

**Comment by adamjstewart** [comment]
Uh, is this something we're planning on doing for all the builders? I thought we created the globals so we wouldn't need all of this cruft.

**Comment by haampie** [comment]
There was a general consensus to get rid of `std_{cmake,meson,pip}_args`, because

1. they are defined in  `build_environment.py` which is the wrong place:
   1. changes to their values is not reflected in the hash.
   2. they are defined for any package, whether they do depend on pip or not.
2. their origin is unclear when just looking at package.py

References to `std_{cmake,meson}_args` were already removed from other PRs.

I don't think `PythonPipBuilder.std_args(self)` is that much cruft?

The alternative is to leave the global, and define them in `{cmake,meson,py-pip}/package.py`
through `setup_dependent_package(child, parent)`, but that is too broad as it should only ever
be called on a build edge `<root spec to be built> - {cmake,meson,py-pip}`, whereas right now
it's called on any in-edge of `{cmake,meson,py-pip}`.

**Comment by adamjstewart** [comment]
Thanks, didn't know there was a discussion on this. As long as there is good reason (like the hash) and we're planning on doing the same for all build systems I'm happy.

**Review by adamjstewart — APPROVED** [review]
Code changes look fine, just need to get CI to pass now.

**Comment by adamjstewart** [comment]
FYI, the [Packaging Guide](https://spack.readthedocs.io/en/latest/packaging_guide.html) still references `std_cmake_args` and may need to be updated with the new syntax.

**Comment by haampie** [comment]
To be fair, the package hash issue is not addressed by this PR because `spack.build_systems.python` is not hashed. The only improvements of this PR are:

1. remove dependencies on inputs from `build_environment.py` which are unaccounted for in the package hash.
2. allow users to find the definition of the default pip args global

Moving the literal value of `std_pip_args` into `py-pip/package.py` would fix the hash issue, but then the question is how to keep `PythonPipBuilder.std_args` in sync with it.

The similar PRs for `std_cmake_args` and `std_meson_args` were much smaller, so let me just bring up this PR in the Monday meeting.

**Comment by haampie** [comment]
@alalazo it looks like the audit is broken:

```
1. Package 'py-pynucleus' need to move the 'setup_build_environment' method from the package class to the appropriate builder class
2. Package 'py-tensorboard-data-server' need to move the 'setup_build_environment' method from the package class to the appropriate builder class
```

**Comment by alalazo** [comment]
For multiple build systems we decided[^1] that the mere presence of a builder class in the `package.py` file was the condition to avoid constructing the wrapper around the package, and use that class as an explicit builder:

https://github.com/spack/spack/blob/11aa02b37adcb04925c75e8746587ea65b5cd812/lib/spack/spack/builder.py#L99-L104

Since in e.g. the `py-pynucleus` package `PipBuilder` is imported by name, the audit detects that, and the build mechanism will use the default builder without the wrapper. I guess, using the fully qualified namespace should not trigger this.

**TL;DR** This seems a hairy edge case, given our current mechanism for builders. Using a fully qualified name should avoid the issue, but wondering if we can improve the implementation to avoid pitfalls like this one.

[^1]: Whether this was a good choice is debatable. I proposed explicit decorators at the time, or other ways to register a custom builder explicitly. The decision went with the proposal that was reducing boilerplate to the minimum in the most common cases, albeit being implicit. This has merits, but also corner cases like this one.

**Comment by haampie** [comment]
Oh... `from x import y` binds `x.y` in the module without it being a definition of `y`, which `getattr` doesn't see  :man_facepalming:

We might as well add an audit that prevents _that_ :grimacing:.

That makes this much uglier than I hoped, cause I don't want to `from spack.build_systems import python` either, since `python` is already used as an executable. And `spack.build_systems.python.PythonPipBuilder.std_args(self)` is awfully long.

Edit: I think this can be prevented by looking at `<builder class>.__module__` and check that it's from `spack.pkg.*`

**Review by jcpunk — APPROVED** [review]

**Comment by alalazo** [comment]
@spackbot run pipeline

Screenshot from 2024-10-29 16-14-10

**Review by rbberger — APPROVED** [review]

**Review by alalazo — APPROVED** [review]
